src/robot/variables/assigner.py

```diff
@@ -134,6 +137,48 @@ def _variable_supports_extended_assign(self, var):
     def _is_valid_extended_attribute(self, attr):
         return self._valid_extended_attr.match(attr) is not None
 
+    def _parse_list_index(self, index):
+        if isinstance(index, (int, slice)):
+            return index
+        if not is_string(index):
+            raise ValueError
+        if ':' not in index:
+            return int(index)
+        if index.count(':') > 2:
+            raise ValueError
+        return slice(*[int(i) if i else None for i in index.split(':')])
+
+    def _variable_type_supports_item_assign(self, var):
+        return (hasattr(var, "__setitem__") and callable(var.__setitem__))
+
+    def _item_assign(self, name, items, value, variables):
+        decorated_items = list(map(lambda item: "[%s]" % item, items))
+        var = variables.replace_scalar("%s%s" % ('$' + name[1:],
+                                                "".join(decorated_items[:-1])))
+
+        if not self._variable_type_supports_item_assign(var):
+            var_type = type_name(var)
+            raise VariableError(
+                f"Variable '{name}{''.join(decorated_items[:-1])}' of type {var_type} "
+                f"does not support item assignment: variable "
+                f"should be mutable and should support a '__setitem__' method."
+                )
+
+        selector = variables.replace_scalar(items[-1])
+
+        try:
+            if isinstance(var, list):
+                selector = self._parse_list_index(selector)
```

Comment:
1. Actually yes `MutableSequence` makes more sense than `list`.
2. I edited the code to try handle `ValueError` from `_parse_sequence_index` (renamed). Now if the parsing fails, the unparsed index is used. I'm not sure in which case the parsing to an `int` or `slice` would fail but then succeed on accessing a `MutableSequence`. But I guess such cases such exist... Anyways I followed your suggestion. ✅

Regarding the test coverage, I believe the atests added in this PR cover the different paths.